lib/rules/element-attributes-order.js

```diff
@@ -0,0 +1,128 @@
+'use strict';
+
+const Rule = require('./base');
+
+module.exports = class ElementAttributesOrder extends Rule {
+  visitor() {
+    const log = this.log.bind(this);
+    const sourceForNode = this.sourceForNode.bind(this);
+
+    function extractElementTokens(node, nodeText) {
+      const keys = [];
+      node.attributes.forEach(attr => {
+        const source = sourceForNode(attr);
+        const type = attr.name.startsWith('@')
+          ? 'arguments'
+          : attr.name.startsWith('...')
+          ? 'splattributes'
+          : 'attributes';
+        keys.push([type, nodeText.indexOf(source), source, attr]);
+      });
+      node.modifiers.forEach(modifier => {
+        const source = sourceForNode(modifier);
+        keys.push(['modifiers', nodeText.indexOf(source), source, modifier]);
+      });
+      return keys;
+    }
+
+    function addMetaToCombinedTokens(tokens) {
+      Object.keys(tokens).forEach(tokenName => {
+        const indexes = tokens[tokenName].map(([index]) => index);
+        tokens[`has_${tokenName}`] = tokens[tokenName].length !== 0;
+        tokens[`${tokenName}_last_index`] = Math.max(indexes);
+        tokens[`${tokenName}_first_index`] = Math.min(indexes);
+      });
+      return tokens;
+    }
+
+    function combineElementTokens(tokens) {
+      const result = {
+        modifiers: [],
+        attributes: [],
+        splattributes: [],
+        arguments: [],
+      };
+      tokens.forEach(([name, index, source, node]) => {
+        result[name].push([index, source, node]);
+      });
+      return result;
+    }
+
+    function checkAttributesOrder(node) {
+      const maybeNodeSource = sourceForNode(node);
+      if (!maybeNodeSource) {
+        return;
+      }
+      const nodeTokens = addMetaToCombinedTokens(
+        combineElementTokens(extractElementTokens(node, maybeNodeSource))
+      );
+      const {
+        arguments_last_index,
+        has_arguments,
+        has_attributes,
+        has_modifiers,
+        has_splattributes,
+        attributes_last_index,
+        attributes_first_index,
+        modifiers_last_index,
+        modifiers_first_index,
+        splattributes_last_index,
+        splattributes_first_index,
+      } = nodeTokens;
+      if (has_arguments) {
+        if (
+          (arguments_last_index > attributes_first_index && has_attributes) ||
```

Comment:
Could you define the default ordering as a constant array? It would make it much easier to follow the code and also enable us to make the ordering configurable in the future.

```js
const ATTRIBUTE_ORDER = ['arguments', 'attributes', 'modifiers', ...];
```

Reply:
should we follow this implementation? https://github.com/vuejs/eslint-plugin-vue/blob/master/lib/rules/attributes-order.js